src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+                lock (_lockObject)
+                {
+                    if (_sessionHandle == null)
+                    {
+                        uint accessType;
+
+                        // If a custom proxy is specified and it is really the system web proxy
+                        // (initial WebRequest.DefaultWebProxy) then we need to update the settings
+                        // since that object is only a sentinel.
+                        if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy)
+                        {
+                            Debug.Assert(state.Proxy != null);
+                            try
+                            {
+                                state.Proxy.GetProxy(state.RequestMessage.RequestUri);
+                            }
+                            catch (PlatformNotSupportedException)
+                            {
+                                // This is the system web proxy.
+                                state.WindowsProxyUsePolicy = WindowsProxyUsePolicy.UseWinInetProxy;
+                                state.Proxy = null;
+                            }
+                        }
+
+                        if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.DoNotUseProxy ||
+                            state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseCustomProxy)
+                        {
+                            // Either no proxy at all or a custom IWebProxy proxy is specified.
+                            // For a custom IWebProxy, we'll need to calculate and set the proxy
+                            // on a per request handle basis using the request Uri.  For now,
+                            // we set the session handle to have no proxy.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_NO_PROXY;
+                        }
+                        else if (state.WindowsProxyUsePolicy == WindowsProxyUsePolicy.UseWinHttpProxy)
+                        {
+                            // Use WinHTTP per-machine proxy settings which are set using the "netsh winhttp" command.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_DEFAULT_PROXY;
+                        }
+                        else
+                        {
+                            // Use WinInet per-user proxy settings.
+                            accessType = Interop.WinHttp.WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY;
+                        }
+
+                        _sessionHandle = Interop.WinHttp.WinHttpOpen(
+                            IntPtr.Zero,
+                            accessType,
+                            Interop.WinHttp.WINHTTP_NO_PROXY_NAME,
+                            Interop.WinHttp.WINHTTP_NO_PROXY_BYPASS,
+                            0);
+                        if (!_sessionHandle.IsInvalid)
+                        {
+                            _sessionHandle.DangerousAddRef(ref ignore);
```

Comment:
@bartonjs I agree the SafeHandle code looks confusing especially regarding the use of the DangerousAddRef/Release calls.  As you noted above, Dispose() actually calls DangerousRelease() which complicates the entire SafeHandle pattern we're trying to use.

The background of the complication is that we need to keep WinHTTP handles alive and "shared" between the WinHttpHandler object and the handoff to the WinHttpResponseStream object.  Once the response stream is alive, the developer can Dispose() of the WinHttpHandler directly or indirectly via disposing the HttpClient object itself.  This causes lifetime issues we're trying to manage.  In addition, we need to have some determinism regarding the lifetime of the WinHTTP handles (session, connection, request).  So, we're trying to use a "COM" based pattern of AddRef/Release using .NET principles of the SafeHandle object.

Our current testing doesn't show any leaks.  We are going to improve/revise this logic when we switch to the using the higher performance "async" pattern for calling WinHTTP APIs.  So, I'm reluctant to change this logic right now but have it logged as a future issue to look over.